### PR TITLE
1843 add height event for bmi

### LINKED ISSUE #1843: BMI: Use height from previous meassurement
The BMI needs weight and height as input parameters. However, the height of a patient isn't usually changing and thus isn't measured every time.

- [ ] store height as a program event
- [ ] if height is not measured in current encounter fall back to last measured height from the program events

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fixes #1843

~~Just putting this in Draft initially as I have a couple of outstanding questions (see notes below).~~

# 👩🏻‍💻 What does this PR do? 
 

Adds a `physicalExaminationHeight` event when a new height is added/saved. This height is then used in subsequent BMI calculations -- unless a new height is added in which case it uses that.

# 🧪 How has/should this change been tested? 

Encounters -> HIV Care -> Physical exam

Has been configured for "BMI" in "HIV Care Program" in latest program schemas repo.

## 💌 Any notes for the reviewer?

- It's not clear to the user how the BMI is being calculated when they haven't entered a height yet. Do you think we should add a "Text" element saying "Previous recorded height: XXX" in the form?
- I'm not sure if the `useProgramEvents.document.list()` query I've added has all the required params. Also, do we want to add an additional query type (other than `.list()`) to just get the latest event of a given type/document/patient? (rather than pulling a list and taking a single value from it?)

### CODE REVIEW

File: client/packages/programs/src/JsonForms/components/BMI.tsx [hunk 1]
```diff
@@ -8,24 +8,60 @@ import {
...
+  const { data: currentEncounter } = useEncounter.document.byId(encounterId);
+  const { data: events } = useProgramEvents.document.list(
+    {
+      filter: {
+        patientId: { equalTo: currentEncounter?.patient?.id ?? '' },
+        type: {
+          equalTo: 'physicalExaminationHeight',
+        },
+        documentType: {
+          equalTo: currentEncounter?.type,
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/BMI.tsx:32]: This needs updating to handle configurable "types"

File: client/packages/programs/src/JsonForms/components/BMI.tsx [hunk 2]
```diff
@@ -8,24 +8,60 @@ import {
...
 import { DetailInputWithLabelRow, NumUtils } from '@openmsupply-client/common';
 import { FORM_LABEL_WIDTH } from '../common';
+import { useEncounter, useProgramEvents } from '../../api';
 
 export const bmiTester = rankWith(10, uiTypeIs('BMI'));
 
 const round = (value: number) => Math.round(value * 100) / 100;
 
+const usePreviousHeight = (
+  formData: any | undefined,
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/BMI.tsx:30]: You could use:

```suggestion
  formData: JsonData | undefined,
```

**Comment by clemens-msupply** [comment]
> 
>     * It's not clear to the user how the BMI is being calculated when they haven't entered a height yet. Do you think we should add a "Text" element saying "Previous recorded height: XXX" in the form?
Yes that would be good. There is already a control for this that displays an event (ProgramEvent control). For example, could do it similar to:

image

and label it "Previous height"?

> 
>     * I'm not sure if the `useProgramEvents.document.list()` query I've added has all the required params. Also, do we want to add an additional query type (other than `.list()`) to just get the latest event of a given type/document/patient? (rather than pulling a list and taking a single value from it?)

Looks good to me.

File: client/packages/programs/src/JsonForms/components/BMI.tsx (hunk 2 above)
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/BMI.tsx]: 👍🏼  f955b02ff

**Comment by CarlosNZ** [comment]
> Yes that would be good. There is already a control for this that displays an event (ProgramEvent control). For example, could do it similar to:

Righto, have added that to the schema:
Screenshot 2023-07-19 at 3 00 38 PM

Would be good if it could be hidden in there was no previous value, is that possible with current options?

**Comment by clemens-msupply** [comment]
> 
> Would be good if it could be hidden in there was no previous value, is that possible with current options?

Thanks for trying it out! Think it looks a bit strange though, to have it in there like that, doesn't it? For example, if you don't know why, it is probably confusing that its there. What do you think about writing the formula out like this:

image

Also very nice would be to have the previous value displayed grayedout in the editable textfield:

image

But this might be more work to implement...

**Comment by CarlosNZ** [comment]
> Also very nice would be to have the previous value displayed grayedout in the editable textfield:
> 
> But this might be more work to implement...

Yeah, "greyed-out" in the Height input would mean you can't edit it.

Could have it in text outside the input, but that would be a different type of form element, not just a simple text input:
254455912-091ff462-c5e7-49aa-87d7-a5a59a43ea54

Might be simpler to just add some more explicit text to the BMI element:
254455912-091ff462-c5e7-49aa-87d7-a5a59a43ea54
(which would only display conditionally)

**Comment by clemens-msupply** [comment]
> Might be simpler to just add some more explicit text to the BMI element: 254455912-091ff462-c5e7-49aa-87d7-a5a59a43ea54 (which would only display conditionally)

looks great and easiest to implement...

**Comment by CarlosNZ** [comment]
> looks great and easiest to implement...
> 
> EDIT: maybe rigth align the text in the edit field along the way?

Cool, that's done now: 10ba2aa60

File: client/packages/programs/src/JsonForms/components/BMI.tsx [hunk 3]
```diff
@@ -6,26 +6,87 @@ import {
...
+        },
+        documentType: {
+          equalTo: currentEncounter?.type,
+        },
+      },
+      page: {
+        first: 1,
+      },
+    },
+    !!currentEncounter && !!eventType
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/BMI.tsx:59]: this enable flag says, if the query is enabled or not, you could only enable it if we need to fetch it e.g.

`const shouldFetchEvent = !formData?.['height'] && !!formData?.['weight'];`

and then:

 `!!currentEncounter && !!eventType && shouldFetchEvent`

File: client/packages/programs/src/JsonForms/components/BMI.tsx [hunk 4]
```diff
@@ -6,26 +6,87 @@ import {
...
+} from '@openmsupply-client/common';
+import { FORM_LABEL_WIDTH, JsonData, useZodOptionsValidation } from '../common';
+import { z } from 'zod';
+import { useEncounter, useProgramEvents } from '../../api';
 
 export const bmiTester = rankWith(10, uiTypeIs('BMI'));
 
 const round = (value: number) => Math.round(value * 100) / 100;
 
+const Options = z
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/BMI.tsx:26]:
```suggestion
/**
 * The BMI component assumes there is a `weight` and `height` property in the current scope and calculates the BMI from these to values.
 * If the `weight` value but not the `height` value is specified, the height value can be looked up from a program event (see `eventType` config).
 */
const Options = z
```

File: client/packages/programs/src/JsonForms/components/BMI.tsx [hunk 5]
```diff
@@ -6,26 +6,87 @@ import {
...
+    /**
+     * Event type to check for "height" value. If not provided, it will always
+     * retrieve from "height" property of current form data.
+     */
+    eventType: z.string().optional(),
+  })
+  .strict();
+type Options = z.infer<typeof Options>;
+
+const usePreviousHeight = (
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/BMI.tsx:37]: nit: just call it `useHeight` because it also gets current height?

**Review by clemens-msupply — CHANGES_REQUESTED** [review]
Look great! just a nitty change request to only fetch when we need to...

File: client/packages/programs/src/JsonForms/components/BMI.tsx (hunk 3 above)
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/BMI.tsx]: Yeah, thanks. fdc7f9efd

(Had to check that `formData` is not undefined too, but basically works)

File: client/packages/programs/src/JsonForms/components/BMI.tsx (hunk 5 above)
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/BMI.tsx]: 029c38e9d

**Comment by CarlosNZ** [comment]
Cool, thanks for the suggestions @clemens-msupply . Working nicely now, I think :)

File: client/packages/programs/src/JsonForms/components/BMI.tsx [hunk 6]
```diff
@@ -6,26 +6,97 @@ import {
...
+import { z } from 'zod';
+import { useEncounter, useProgramEvents } from '../../api';
 
 export const bmiTester = rankWith(10, uiTypeIs('BMI'));
 
 const round = (value: number) => Math.round(value * 100) / 100;
 
+/**
+ * The BMI component assumes there is a `weight` and `height` property in the
+ * current scope and calculates the BMI from these to values. If the `weight`
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/BMI.tsx:28]:
```suggestion
 * current scope and calculates the BMI from these two values. If the `weight`
```

File: client/packages/programs/src/JsonForms/components/BMI.tsx [hunk 7]
```diff
@@ -36,13 +107,28 @@ const UIComponent = (props: ControlProps) => {
...
+          // gap={FORM_GAP}
+        >
+          <BasicTextInput
+            disabled
+            value={data?.bodyMassIndex ?? ''}
+            sx={{ margin: 0.5, width: '90px' }}
+            textAlign="right"
+          />
+          {weight && source === 'previous' && (
+            <Typography sx={{ maxWidth: 150, fontSize: '75%' }}>
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/BMI.tsx:126]: Why didn't you let it go full witdth, e.g. to just have one line?

**Review by clemens-msupply — APPROVED** [review]
Works great thank you!

**Pull request merged**